feature: Add an option to enable Telegram notifications to work behind a proxy.

MatanAmoyal1 opened the pull request:
## 📋 Overview

- **What problem does this pull request address?**
  - Add an option to enable Telegram notifications to work behind a proxy. This allows users to specify proxy settings for outgoing Telegram API requests, ensuring notifications function correctly in proxied network environments.
- **What features or functionality does this pull request introduce or enhance?**
  - Option to enable Telegram notifications to work behind a proxy.

## 🛠️ Type of change

<!-- Please select all options that apply -->

- [ ] 🐛 Bugfix (a non-breaking change that resolves an issue)
- [x] ✨ New feature (a non-breaking change that adds new functionality)
- [ ] ⚠️ Breaking change (a fix or feature that alters existing functionality in a way that could cause issues)
- [ ] 🎨 User Interface (UI) updates
- [ ] 📄 New Documentation (addition of new documentation)
- [ ] 📄 Documentation Update (modification of existing documentation)
- [ ] 📄 Documentation Update Required (the change requires updates to related documentation)
- [ ] 🔧 Other (please specify):
  - Provide additional details here.

## 📄 Checklist

<!-- Please select all options that apply -->

- [x] 🔍 My code adheres to the style guidelines of this project.
- [x] 🦿 I have indicated where (if any) I used an LLM for the contributions
- [x] ✅ I ran ESLint and other code linters for modified files.
- [x] 🛠️ I have reviewed and tested my code.
- [x] 📝 I have commented my code, especially in hard-to-understand areas (e.g., using JSDoc for methods).
- [x] ⚠️ My changes generate no new warnings.
- [x] 🤖 My code needed automated testing. I have added them (this is an optional task).
- [x] 📄 Documentation updates are included (if applicable).
- [x] 🔒 I have considered potential security impacts and mitigated risks.
- [x] 🧰 Dependency updates are listed and explained.
- [x] 📚 I have read and understood the [Pull Request guidelines](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#recommended-pull-request-guideline).

oelezra commented:
This looks great! This is exactly what I need to upgrade my Uptime Kuma. I'll be waiting for this fix.

CommanderStorm reviewed (changes requested):
I would like for all notification providers to behave roughly the same way.
=> I think this code should be moved to the part all notification providers share to make this simpler.

One notification provider being different is just confusing

MatanAmoyal1 commented:
Thanks @CommanderStorm , this is a great idea.
I moved the function to be in the shared class between the notification providers.
And I implemented it only in Telegram, because I was able to verify this only there.

CommanderStorm commented on server/notification-providers/notification-provider.js:
```diff
@@ -115,6 +116,22 @@ class NotificationProvider {
+     * @returns {object} Axios config
+     */
+    getAxiosConfigWithProxy(axiosConfig = {}) {
+        const proxyEnv = process.env.https_proxy || process.env.HTTPS_PROXY || process.env.http_proxy || process.env.HTTP_PROXY;
```
Given the actual rollout, I am a bit concerned.
I don't want to affect users notifications this late in the v2.0 cycle.

We need to move this to a different environment variable or add a setting.

CommanderStorm replied:
I won't merge it in the v2.0 cycle as is, this is too risky.
If you insist on this approach, we can close and revisit this once v3.0 gets going.

MatanAmoyal1 replied:
What is the alternative you suggest to push it to v2 cycle?
Can I use an env variable, for example: $USE_NOTIFICATIONS_PROXY=true, and then if it is set and has a true value, it will use the http_proxy env variables?
In this case it will be strongly protected, and only users with the combination of http_proxy and USE_NOTIFICATIONS_PROXY environment variables will have this proxy?

CommanderStorm commented on server/notification-providers/telegram.js:
```diff
@@ -30,9 +30,9 @@ class Telegram extends NotificationProvider {
-            await axios.get(`${url}/bot${notification.telegramBotToken}/sendMessage`, {
-                params: params,
-            });
+            let axiosConfig = this.getAxiosConfigWithProxy({ params });
+
+            await axios.get(`${url}/bot${notification.telegramBotToken}/sendMessage`, axiosConfig);
```
Let's roll out this change across all notifications. This should be easy enough and I don't see how this can cause harm.

I don't think this needs testing if it works on one notification provider..

CommanderStorm reviewed (changes requested):
Much nicer, I am still a bit concerned about how we roll this out and how this change splits the notifications even more than they already are

MatanAmoyal1 commented:
@CommanderStorm I added the new env variable and implemented it on all notification providers.

CommanderStorm commented on server/notification-providers/flashduty.js:
```diff
@@ -90,6 +90,7 @@ class FlashDuty extends NotificationProvider {
             options.client_url = baseURL + getMonitorRelativeURL(monitorInfo.id);
         }
 
+        options = this.getAxiosConfigWithProxy(options);
```
I am not sure if this works the same way.

Because I cannot test this integration, let's either (preferred) move this to the `axios.post(url, data, config)` API that we are using elsewhere or remove the proxy feature from these notification providers.

CommanderStorm commented on server/notification-providers/gorush.js:
```diff
@@ -32,7 +32,7 @@ class Gorush extends NotificationProvider {
                 ]
             };
             let config = {};
-
+            config = this.getAxiosConfigWithProxy(config);
```
```suggestion
            const config = this.getAxiosConfigWithProxy({});
```

CommanderStorm commented on server/notification-providers/notification-provider.js:
```diff
@@ -115,6 +116,22 @@ class NotificationProvider {
+     * @returns {object} Axios config
+     */
+    getAxiosConfigWithProxy(axiosConfig = {}) {
+        if (process.env.NOTIFICATION_PROXY === "true") {
+            const proxyEnv = process.env.https_proxy || process.env.HTTPS_PROXY || process.env.http_proxy || process.env.HTTP_PROXY;
```
I was more thinking about something like this.
What do you think?

```suggestion
        const proxyEnv = process.env.notification_proxy || process.env.NOTIFICATION_PROXY;
        if (proxyEnv) {
```

CommanderStorm commented on server/notification-providers/ntfy.js:
```diff
@@ -22,6 +22,7 @@ class Ntfy extends NotificationProvider {
                     "Authorization": "Bearer " + notification.ntfyaccesstoken,
                 };
             }
+            let config = this.getAxiosConfigWithProxy({ headers: headers });
```
```suggestion
            let config = { headers: headers };
            config = this.getAxiosConfigWithProxy(config);
```

CommanderStorm reviewed (changes requested):
Nice work.

I have one minor (simple to apply) change request regarding `options` and one api-design comment, but otherwise LGTM

MatanAmoyal1 commented:
> Thanks a ton for all of this work 🎉

Thank you for the review and the guidance :)

louislam commented:
Hi, I think it needs some follow-up, I received a critical alert after this pull request got merged:

<img width="1566" height="446" alt="image" src="https://github.com/user-attachments/assets/746ba465-fa2f-4fa0-bc4c-59a030d9674d" />

MatanAmoyal1 commented:
> Hi, I think it needs some follow-up, I received a critical alert after this pull request got merged:
> <img alt="image" width="1566" height="446" src="https://private-user-images.githubusercontent.com/1336778/495813880-746ba465-fa2f-4fa0-bc4c-59a030d9674d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTkyNjA5NzEsIm5iZiI6MTc1OTI2MDY3MSwicGF0aCI6Ii8xMzM2Nzc4LzQ5NTgxMzg4MC03NDZiYTQ2NS1mYTJmLTRmYTAtYmM0Yy01OWEwMzBkOTY3NGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDkzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA5MzBUMTkzMTExWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDE5M2UxYzk5MWUzNzVjMGRiYmZiMTc1MTY5MjlhNjA2ZDc1ODlhYTA1Yjc5YzQ1YzBhMDBjMDY5OTY1NDY3YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.s1Vn44hIKTy7FW4FD1c_QdRlYFMjqdFcy_KhVS7LqBI">

I will check how to solve it

MatanAmoyal1 commented:
> Hi, I think it needs some follow-up, I received a critical alert after this pull request got merged:
> <img alt="image" width="1566" height="446" src="https://private-user-images.githubusercontent.com/1336778/495813880-746ba465-fa2f-4fa0-bc4c-59a030d9674d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTkzMDI0ODcsIm5iZiI6MTc1OTMwMjE4NywicGF0aCI6Ii8xMzM2Nzc4LzQ5NTgxMzg4MC03NDZiYTQ2NS1mYTJmLTRmYTAtYmM0Yy01OWEwMzBkOTY3NGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTAwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEwMDFUMDcwMzA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGVjZjAzZTUwM2NiM2ZkNzQwYjQ3NDQ2OWVmYjZkNzUzZWM4ODBlNzQ0NjFmZjUyMWU3NzVhMWM5ZWMyMDcwMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.A-9HShhZfx9OWxWdzBH5X4s8X11vqAMjK_gD9hmkDhs">

please review
https://github.com/louislam/uptime-kuma/pull/6156
